tools/graph-convert/graph-convert.cpp

```diff
@@ -2868,18 +2875,90 @@ AppendEdgeData<void>(
 
 /**
  * Gr2Kg reads in the binary csr (.gr) files and produces
- * katana graph property graphs. 
+ * katana graph property graphs.
  */
 struct Gr2Kg : public Conversion {
+  katana::Result<void> OutOfCoreConvert(
+      const std::string& in_file_name, const std::string& out_file_name) {
+    tsuba::GRHeader header;
+    if (auto res = tsuba::FileGet(in_file_name, &header); !res) {
+      return res.error();
+    }
+
+    if (header.version != 1) {
+      KATANA_LOG_ERROR("Out of core not possible, katana expects GR v1");
+      return katana::ErrorCode::NotImplemented;
+    }
+
+    if (header.edge_type_size != 0) {
+      KATANA_LOG_WARN("ignoring existing edge property in conversion");
+      header.edge_type_size = 0;
+    }
+
+    tsuba::StatBuf stat_buf;
+    if (auto res = tsuba::FileStat(in_file_name, &stat_buf); !res) {
+      KATANA_LOG_DEBUG("could not stat {}", out_file_name);
+      return res.error();
+    }
+
+    uint64_t new_size = tsuba::GRFileSize(header);
+    if (stat_buf.size < new_size) {
+      KATANA_LOG_ERROR(
+          "{} does not appear to be well formed (too small)", in_file_name);
+      return katana::ErrorCode::InvalidArgument;
+    }
+
+    if (auto res = tsuba::Create(out_file_name); !res) {
+      return res.error();
+    }
+
+    auto handle_res = tsuba::Open(out_file_name, tsuba::kReadWrite);
+    if (!handle_res) {
+      return handle_res.error();
+    }
+
+    tsuba::RDGFile handle(std::move(handle_res.value()));
+
+    katana::Uri top_file_name = tsuba::NewTopologyFileName(handle);
+    if (auto res =
+            tsuba::FileCopy(in_file_name, top_file_name.string(), 0, new_size);
+        !res) {
+      return res.error();
+    }
+
+    tsuba::RDG rdg;
+    if (auto res = rdg.RegisterTopologyFile(top_file_name); !res) {
+      return res.error();
+    }
+    return rdg.Store(handle, kCommandLine);
+  }
+
   template <typename EdgeTy>
-  void convert(const std::string& infilename, const std::string& outfilename) {
+  void convert(
+      const std::string& in_file_name, const std::string& out_file_name) {
     using Graph = katana::FileGraph;
     using GNode = Graph::GraphNode;
     using EdgeData = katana::LargeArray<EdgeTy>;
     using edge_value_type = typename EdgeData::value_type;
 
+    if constexpr (std::is_same<EdgeTy, void>::value) {
+      // the property graph topology file format is very close to gr, so we can
+      // use this shortcut. This shortcut also avoids reading the graph on one
+      // host so it's important to do this to support large graphs
+      fmt::print(stderr, "attempting out-of-core conversion\n");
```

Comment:
for later: I'm wondering if we should add some callback/status API so that callers can implement some progress tracking / monitoring on expensive operations like partitioning and loading, e.g., `ExpensiveOperation(....., [](size_t n, size_t total) { print("did {n}th of {total}"); })`

Reply:
Seems reasonable to me; an alternative would be to enrich the logs and tools for parsing them in real time with something like `prometheus`?

Reply:
Or I guess call backs like this are general enough to be the mechanism that we would use to add prometheus/metrics support.

Reply:
Yeah; users can do whatever they want and the core library doesn't need to take on any specific logging library as a dependency.

If/when we go down this route, I have some preference for a pretty generic callback function: `ExpensiveOperation(..., [](FooContext ctx) { })` as that will give us some flexibility about the actual parameters passed and give the callee a handle to query the system if necessary.